File: app/ruby/src/examples/aoc.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# laurent.martin.aspera@fr.ibm.com
# Aspera on Cloud
# Send a package to shared inbox (name in config file) in given workspace (name in config file)

require_relative '../utils/configuration'
require_relative '../utils/transfer_client'
require_relative '../utils/rest'
require 'logger'
require 'base64'
require 'securerandom'

AOC_API_V1_BASE_URL = 'https://api.ibmaspera.com/api/v1'
AOC_OAUTH_AUDIENCE = 'https://api.asperafiles.com/api/v1/oauth2/token'

PACKAGE_NAME        = 'sample package Ruby'
TRANSFER_SESSIONS   = 1

config = Utils::Configuration.instance
transfer_client = Utils::TransferClient.new(config).startup

# Generate a transfer cookie for AoC
def aoc_xfer_cookie(app, user_name, user_id)
  encoded_app       = Base64.strict_encode64(app)
  encoded_user_name = Base64.strict_encode64(user_name)
  encoded_user_id   = Base64.strict_encode64(user_id)
  "aspera.aoc:#{encoded_app}:#{encoded_user_name}:#{encoded_user_id}"
end

# Generate transfer spec for gen4 API, in simplified way
def gen4_base_spec(
  aoc_api,
  app,
  dir,
  node_info,
  user_info,
  workspace_info,
  app_info
)
  op = dir.eql?('send') ? 'upload' : 'download'
  s = {
    'direction' => dir,
    'remote_host' => node_info['host'],
    'remote_user' => 'xfer',
    'ssh_port' => 33_001,
    'fasp_port' => 33_001,
    'cookie' => aoc_xfer_cookie(app, user_info['name'], user_info['email']),
    'create_dir' => true,
    'target_rate_kbps' => 300_000,
    'token' => aoc_api.bearer_token_authorization("node.#{node_info['access_key']}:user:all"),
    'tags' => {
      'aspera' => {
        'app' => app,
        'usage_id' => "aspera.files.workspace.#{workspace_info['id']}",
        'files' => {
          'node_id' => node_info['id'],
          'workspace_name' => workspace_info['name'],
          'workspace_id' => workspace_info['id'],
          'files_transfer_action' => "#{op}_#{app.gsub(/s$/, '')}"
        },
        'node' => {
          'access_key' => node_info['access_key']
        },
        'xfer_retry' => 3600
      }
    }
  }
  astags = s['tags']['aspera']
  case app
  when 'packages'
    # app_info is the package information
    astags['files']['package_id'] = app_info['id']
    astags['files']['package_name'] = app_info['name']
    astags['files']['package_operation'] = op
    astags['node']['file_id'] = app_info['contents_file_id']
  when 'files'
    # app_info is the id of the folder
    astags['node']['file_id'] = app_info
    astags['files']['parentCwd'] = "#{node_info['id']}:#{app_info}"
  end
  s
end

begin
  logger = config.logger
  aoc_api = Utils::Rest.new(AOC_API_V1_BASE_URL)
  aoc_api.auth_bearer(
    token_url: "#{AOC_API_V1_BASE_URL}/oauth2/#{config.param('aoc', 'org')}/token",
    key_pem_path: config.param('aoc', 'private_key'),
    client_id: config.param('aoc', 'client_id'),
    client_secret: config.param('aoc', 'client_secret'),
    iss: config.param('aoc', 'client_id'),
    aud: AOC_OAUTH_AUDIENCE,
    sub: config.param('aoc', 'user_email'),
    org: config.param('aoc', 'org')
  )
  aoc_api.default_scope('user:all')

  # Get my user information
  user_info = aoc_api.read('self')
  logger.debug(user_info)

  # Get workspace information
  workspace_name = config.param('aoc', 'workspace')
  logger.info("getting workspace information for #{workspace_name}")
  response_data = aoc_api.read('workspaces', { 'q' => workspace_name })
  logger.debug(response_data)
  raise "Found #{response_data.size} workspace(s) for #{workspace_name}" unless response_data.size == 1

  workspace_info = response_data.first

  #---------------
  # Packages
  #===============

  # Get shared inbox information
  shared_inbox_name = config.param('aoc', 'shared_inbox')
  logger.info('getting shared inbox information')
  response_data = aoc_api.read(
    'dropboxes',
    {
      'current_workspace_id' => workspace_info['id'],
      'q' => shared_inbox_name
    }
  )
  logger.debug(response_data)
  raise "Found #{response_data.size} dropbox for #{shared_inbox_name}" unless response_data.size == 1

  dropbox_info = response_data.first

  # Create a new package
  logger.info('creating package')
  package_info = aoc_api.create(
    'packages', {
      'workspace_id' => workspace_info['id'],
      'recipients' => [{ 'id' => dropbox_info['id'], 'type' => 'dropbox' }],
      'name' => PACKAGE_NAME,
      'note' => 'My package note',
      'sent' => true,
      'transfers_expected' => TRANSFER_SESSIONS
    }
  )
  logger.debug(package_info)

  # Get node information
  logger.info('getting node information')
  package_node_info = aoc_api.read("nodes/#{package_info['node_id']}")
  logger.debug(package_node_info)

  # Build transfer spec
  t_spec = gen4_base_spec(
    aoc_api,
    'packages',
    'send',
    package_node_info,
    user_info,
    workspace_info,
    package_info
  )

  if TRANSFER_SESSIONS != 1
    t_spec['multi_session']           = TRANSFER_SESSIONS
    t_spec['multi_session_threshold'] = 500_000
  end

  # Add file list
  config.add_sources(t_spec, 'paths')

  # Start transfer
  transfer_client.start_transfer_and_wait(t_spec)

  #---------------
  # Files
  #===============

  home_node_info = aoc_api.read("nodes/#{workspace_info['home_node_id']}")
  logger.debug(home_node_info)

  # Upload to home in workspace (Files app)
  transfer_spec = gen4_base_spec(
    aoc_api,
    'files',
    'send',
    home_node_info,
    user_info,
    workspace_info,
    workspace_info['home_file_id'] # upload directly into home
  )

  # Add list of files to upload
  config.add_sources(transfer_spec, 'paths')
  logger.info("spec: #{transfer_spec}")

  # Start transfer
  transfer_client.start_transfer_and_wait(transfer_spec)
ensure
  transfer_client.shutdown
end

File: app/rust/src/examples/aoc.rs
use log::info;
use samples::utils::configuration::Configuration;
use samples::utils::rest;
use samples::utils::transfer_client::TransferClient;
use serde_json::json;
use std::error::Error;
use std::sync::Arc;

const AOC_API_V1_BASE_URL: &str = "https://api.ibmaspera.com/api/v1";
const AOC_OAUTH_AUDIENCE: &str = "https://api.asperafiles.com/api/v1/oauth2/token";

/// Example showing access to Aspera on Cloud (AoC) API
#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Configuration and Transfer Client setup
    let config = Arc::new(Configuration::new()?);
    let mut transfer_client: TransferClient = TransferClient::new(Arc::clone(&config));

    // Initialize the AoC API REST client
    let mut aoc_api = rest::Rest::new(AOC_API_V1_BASE_URL, true)?;

    // Set Bearer token parameters for authentication using JWT
    aoc_api.set_bearer(rest::BearerData {
        token_url: format!(
            "{}/oauth2/{}/token",
            AOC_API_V1_BASE_URL,
            config.param_str("aoc", "org")?
        ),
        key_pem_path: config.param_str("aoc", "private_key")?,
        client_id: config.param_str("aoc", "client_id")?,
        client_secret: config.param_str("aoc", "client_secret")?,
        iss: config.param_str("aoc", "client_id")?,
        aud: AOC_OAUTH_AUDIENCE.to_string(),
        sub: config.param_str("aoc", "user_email")?,
        org: Some(config.param_str("aoc", "org")?),
    });
    // generate the bearer token
    aoc_api
        .set_default_scope(Some("user:all".to_string()))
        .await?;

    // Get workspace information
    let workspace_name = config.param_str("aoc", "workspace")?;
    info!("Getting workspace information for {}", workspace_name);

    let workspace_response = aoc_api
        .read("workspaces", Some(&[("q", &workspace_name)]))
        .await?;
    let workspace_list = workspace_response.as_array().unwrap();
    if workspace_list.len() != 1 {
        return Err(Box::from(format!(
            "Found {} workspaces for {}",
            workspace_list.len(),
            workspace_name
        )));
    }
    let workspace_info = &workspace_list[0];

    // Get shared inbox (dropbox) information
    let shared_inbox_name = config.param_str("aoc", "shared_inbox")?;
    info!("Getting shared inbox information");
    let dropbox_response = aoc_api
        .read(
            "dropboxes",
            Some(&[
                ("current_workspace_id", &workspace_info["id"].to_string()),
                ("q", &shared_inbox_name),
            ]),
        )
        .await?;
    let dropbox_list = dropbox_response.as_array().unwrap();
    if dropbox_list.len() != 1 {
        return Err(Box::from(format!(
            "Found {} dropboxes for {}",
            dropbox_list.len(),
            shared_inbox_name
        )));
    }
    let dropbox_info = &dropbox_list[0];

    // Create a new package
    info!("Creating package");
    let package_response = aoc_api
        .create(
            "packages",
            &json!({
                "workspace_id": workspace_info["id"],
                "recipients": [{ "id": dropbox_info["id"], "type": "dropbox" }],
                "name": "sample package",
                "note": "My package note"
            }),
            None,
        )
        .await?;
    let package_info = package_response.as_object().unwrap();
    info!("Package created: {:?}", package_info);

    // Get node information for the package
    let node_id = package_info["node_id"].as_str().unwrap();
    info!("Getting node information for {node_id}");
    let node_response = aoc_api.read(&format!("nodes/{node_id}"), None).await?;
    let node_info = node_response.as_object().unwrap();
    info!("Node information: {:?}", node_info);

    // Set expected transfers for the package
    info!("Setting expected transfers");
    aoc_api
        .update(
            &format!("packages/{}", package_info["id"].as_str().unwrap()),
            &json!({
                "sent": true,
                "transfers_expected": 1
            }),
        )
        .await?;

    info!("Creating Transfer spec");
    // Generate the transfer spec
    let mut transfer_spec = json!({
        "direction": "send",
        "token": aoc_api.get_bearer_token(Some(format!("node.{}:user:all", node_info["access_key"].as_str().unwrap()))).await?,
        "tags": {
            "aspera": {
                "app": "packages",
                "files": {
                    "node_id": node_info["id"].as_str().unwrap(),
                    "package_id": package_info["id"].as_str().unwrap(),
                    "package_name": package_info["name"].as_str().unwrap(),
                    "package_operation": "upload",
                    "files_transfer_action": "upload_package",
                    "workspace_name": workspace_info["name"].as_str().unwrap(),
                    "workspace_id": workspace_info["id"].as_str().unwrap()
                },
                "node": {
                    "access_key": node_info["access_key"].as_str().unwrap(),
                    "file_id": package_info["contents_file_id"].as_str().unwrap()
                },
                "usage_id": format!("aspera.files.workspace.{}", workspace_info["id"].as_str().unwrap()),
                "xfer_id": uuid::Uuid::new_v4().to_string(),
                "xfer_retry": 3600
            }
        },
        "remote_host": node_info["host"].as_str().unwrap(),
        "remote_user": "xfer",
        "ssh_port": 33001,
        "fasp_port": 33001,
        "create_dir": true,
        "target_rate_kbps": 100000,
        "paths": []
    });

    // Add files to transfer spec
    config.add_files_to_ts("paths", &mut transfer_spec)?;

    // Upload files to package folder on server
    transfer_client
        .transfer_start_and_wait(&transfer_spec)
        .await?;

    Ok(())
}
